### PR TITLE
Fix mac applet problems

### LINKED ISSUE #13246: mac util requires 2 EOFs
When using "openssl mac", it requires to see 2 EOFs before it calculates the output.

### PR DESCRIPTION
**Pull request by paulidale**
This fixes the spurious newline problem and the double EOF problem.
It also adds -cipher and -digest command line options to make life a little easier.

Fixes #13246 and #13247

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: apps/mac.c [hunk 1]
```diff
@@ -48,6 +52,27 @@ const OPTIONS mac_options[] = {
     {NULL}
 };
 
+static char *alloc_mac_algorithm_name(STACK_OF(OPENSSL_STRING) **optp,
+                                      const char *name, const char *arg)
+{
+    size_t len = strlen(name) + strlen(arg) + 2;
+    char *res = app_malloc(len, "algorithm name");
```
Comment (slontis) [inline apps/mac.c:59]: oops - double malloc..

File: apps/mac.c [hunk 2]
```diff
@@ -93,6 +119,16 @@ int mac_main(int argc, char **argv)
             if (opts == NULL || !sk_OPENSSL_STRING_push(opts, opt_arg()))
                 goto opthelp;
             break;
+        case OPT_CIPHER:
+            cipher = alloc_mac_algorithm_name(&opts, "cipher", opt_arg());
```
Comment (slontis) [inline apps/mac.c:123]: This looks like a slippery slope to me :)

File: apps/mac.c (hunk 1 above)
Comment (paulidale) [reply apps/mac.c:59]: I just spent the last half hour trying to locate this and was about to give up :(

File: doc/man1/openssl-mac.pod.in [hunk 3]
```diff
@@ -66,20 +82,6 @@ Specifies the MAC key in hexadecimal form (two hex digits per byte).
 The key length must conform to any restrictions of the MAC algorithm.
 A key must be specified for every MAC algorithm.
 
-=item B<digest:>I<string>
```
Comment (slontis) [inline doc/man1/openssl-mac.pod.in:69]: These options havent really disappeared from here :)

File: apps/mac.c (hunk 2 above)
Comment (paulidale) [reply apps/mac.c:123]: Wheeee!!!

File: doc/man1/openssl-mac.pod.in (hunk 3 above)
Comment (paulidale) [reply doc/man1/openssl-mac.pod.in:69]: Look further down :)  I moved them to the end.

File: apps/mac.c (hunk 1 above)
Comment (slontis) [reply apps/mac.c]: Most people do double frees - you are special...

File: apps/mac.c [hunk 4]
```diff
@@ -93,6 +116,16 @@ int mac_main(int argc, char **argv)
...
+            if (cipher == NULL)
+                goto opthelp;
+            break;
+        case OPT_DIGEST:
+            digest = alloc_mac_algorithm_name(&opts, "digest", opt_arg());
+            if (digest == NULL)
+                goto opthelp;
+            break;
```
Comment (t8m) [inline apps/mac.c:128]: if the options are used multiple times, it will leak memory

Comment (paulidale) [reply apps/mac.c]: Fixed.

**Review by t8m — APPROVED** [review]

**Comment by richsalz** [comment]
It's not an `applet`  It's a `command`
:)

**Comment by paulidale** [comment]
Merged to master, thanks for the reviews.

@richsalz, I never was good with names :)

**Pull request closed without merging**